drivers: platform: Add SPI support for ADuCM3029

mchindri opened the pull request:
Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=335) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=335) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=335) it.</sub>

mchindri commented:
Deleted spaces and modified commit description.

mchindri commented:
v3
 - Moved attribution after checking pointer for NULL value.

mirceacaprioru commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+	/* Verify if the SPI is already activated */
+	adicup_init_param = param->extra;
+	if (active_spi & (1 << adicup_init_param->spi_channel))
+		goto fail_mem;
```
you don't have memory allocated yet.

mchindri replied:
_param_ is allocated by the user and _adicup_init_param_ is used only for readability to avoid long casting

mirceacaprioru commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+
+	*desc = alloc_data_mem(param);
+	if (*desc == NULL)
+		goto fail_no_mem;
```
return Failure;

mirceacaprioru commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+	struct spi_desc		*desc;
+	struct adicup_spi_desc	*extra;
+
+	desc = malloc(sizeof(struct spi_desc));
```
calloc would be a better option since it initializes the memory to 0.

mchindri replied:
Every malloc or only this one?

amiclaus replied:
You can have a look at latest platform specific drivers that are merged: https://github.com/analogdevicesinc/no-OS/blob/724c8dad5d1aedec46d1ce56c4b7fbeabe3504f2/drivers/platform/xilinx/spi.c#L59-L78

Generally, `calloc` is preferred over `malloc` since it is more risk-free.

scuciurean commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+	/* SPI configured. Set it as activated */
+	active_spi |= (1 << adicup_init_param->spi_channel);
+
+	spi_ret = SUCCESS;
```
You can return SUCCESS at this point.

scuciurean commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+	free_desc_mem(*desc);
+fail_no_mem:
+	spi_ret = FAILURE;
+succes:
```
The label is a bit confusing since you go there even if the return value is FAILURE

mchindri commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+	struct adicup_spi_desc	*extra;
+
+	desc = malloc(sizeof(struct spi_desc));
+	if (NULL == desc)
```
alloc_data_mem is like a malloc function that's why I used the return NULL model.

amiclaus commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,232 @@
+struct spi_desc *alloc_data_mem()
+{
+	struct spi_desc		*desc;
+	struct adicup_spi_desc	*extra;
```
I suggest keeping the variable naming consistent with other drivers.
`*extra` -> `*adicup_desc`
I.e.
https://github.com/analogdevicesinc/no-OS/blob/724c8dad5d1aedec46d1ce56c4b7fbeabe3504f2/drivers/platform/xilinx/spi.c#L62-L65

mchindri commented:
V4
 - Retruned directly from spi_init when possible and renamed failure label.
 - Changed comparation with NULL with !
 - Changed malloc with calloc
 - Rename the local variable extra to adicup_desc

adrimbarean commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,228 @@
+ * @brief Allocates the memory needed for the descriptor
+ * @return A address to the allocated memory, NULL if the allocation fails.
+ */
+struct spi_desc *alloc_data_mem(void)
```
nitpick: this function can be static if it's not intended to be an API.

adrimbarean commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,228 @@
+ * @brief Deallocates the memory needed for the descriptor
+ * @param desc - Descriptor to be deallocated
+ */
+void free_desc_mem(struct spi_desc *desc)
```
nitpick: this function can be static if it's not intended to be an API.

adrimbarean commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,228 @@
+
+	/* Verify if the SPI is already activated */
+	adicup_init_param = param->extra;
+	if (active_spi & (1 << adicup_init_param->spi_channel))
```
`adi_spi_Open` returns specific error code in case the spi channel is already in use. Do we still need this check considering this?

mchindri replied:
Adi_spi_open returns already in use only in debug mode, so it is better to keep it

mchindri commented:
V5
 - Make _free_desc_mem_ and _alloc_data_mem_ static

amiclaus commented on drivers/platform/aducm3029/spi.c:
```diff
@@ -0,0 +1,228 @@
+ */
+static void free_desc_mem(struct spi_desc *desc)
+{
+	free(((struct adicup_spi_desc*)(desc->extra))->buffer);
```
I would check here for each pointer if deallocation is required.

Maybe use `adicup_spi_desc` type pointer to avoid casting.

mchindri commented:
V6
- Modify description of the pull request
- Comment structures from spi_extra.h
- Update type of bytes_number from uint8_t to uint16_t

cristiantpop commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,71 @@
+	MASTER										// SPI used in Slave mode
+};
+
+struct adicup_spi_init_param {
```
Please update comments to structs like here:
https://github.com/analogdevicesinc/no-OS/commit/1545b6e21c5d9ccb185a6692e70de68bb9038f52

mchindri commented:
V7
- Updated structure comments like https://github.com/analogdevicesinc/no-OS/commit/1545b6e21c5d9ccb185a6692e70de68bb9038f52
- Modified structure names from adicup to aducm

amiclaus commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,101 @@
+ * @half_duplex: If true, it enables half duplex mode. The default if false
+ * @dma: Device name. If true, it enables DMA transfer. The default value is false
+ * @buffer:	RESERVED
+ * @spi_handle:	Handle to identify the SPI device
```
Looks like you are using combination of tabs and spaces here.

amiclaus commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,101 @@
+	ADI_SPI_HANDLE			spi_handle;
+};
+
+
```
nitpick: extra line.

amiclaus commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,101 @@
+	MASTER
+};
+
+
```
nitpick: extra line.

amiclaus commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,101 @@
+	SPI2
+};
+
+
```
nitpick: extra line.

amiclaus commented on drivers/platform/aducm3029/spi_extra.h:
```diff
@@ -0,0 +1,101 @@
+#ifndef SPI_EXTRA_H
+# define SPI_EXTRA_H
+
+#include <drivers/spi/adi_spi.h>
```
add please separation comment: "include files"
https://github.com/analogdevicesinc/no-OS/blob/bf5962d288dc7e98836820ac465f53034c826997/drivers/platform/xilinx/spi_extra.h#L42-L46

amiclaus replied:
and "types declaration":
https://github.com/analogdevicesinc/no-OS/blob/bf5962d288dc7e98836820ac465f53034c826997/drivers/platform/xilinx/spi_extra.h#L54-L56

mchindri commented:
V8
- Deleted extra empty lines
- Added Types Declarations and Include Files comment
- Changed spaces with tabs in comments

mchindri commented:
V9
- Check in function spi_write_and_read if bytes_number is 0 and return FAILURE. Added this because the SPI gets stuck if this parameter is 0.
